x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/TomlFoldTests.java

```diff
@@ -0,0 +1,118 @@
+
+import static java.util.stream.Collectors.toList;
+
+public class TomlFoldTests extends ESTestCase {
+    protected static final String PARAM_FORMATTING = "%1$s.test -> %2$s";
+
+    @ParametersFactory(shuffle = false, argumentFormatting = PARAM_FORMATTING)
+    public static List<Object[]> readTestSpecs() throws Exception {
+        List<EqlFoldSpec> foldSpecs = EqlFoldSpecLoader.load("/test_folding.toml");
+        foldSpecs.addAll(EqlFoldSpecLoader.load("/test_string_functions.toml"));
+        List<EqlFoldSpec> unsupportedSpecs = EqlFoldSpecLoader.load("/test_unsupported.toml");
+
+        HashSet<EqlFoldSpec> filteredSpecs = new HashSet<>(foldSpecs);
+        filteredSpecs.removeAll(unsupportedSpecs);
+        return asArray(filteredSpecs);
+    }
+
+    public static List<Object[]> asArray(Collection<EqlFoldSpec> specs) {
+        AtomicInteger counter = new AtomicInteger();
+        return specs.stream().map(spec -> new Object[] {
+            counter.incrementAndGet(), spec
+        }).collect(toList());
+    }
+
+    private static final EqlConfiguration caseSensitiveConfig = config(true);
+    private static final EqlConfiguration caseInsensitiveConfig = config(false);
+    private static EqlParser parser = new EqlParser();
+    private static final EqlFunctionRegistry functionRegistry = new EqlFunctionRegistry();
+
+    private final int num;
+    private final EqlFoldSpec spec;
+
+    private static EqlConfiguration config(boolean isCaseSensitive) {
+        return new EqlConfiguration(new String[]{"none"},
+            org.elasticsearch.xpack.ql.util.DateUtils.UTC, "nobody", "cluster",
+            null, TimeValue.timeValueSeconds(30), -1, false, isCaseSensitive, "",
+            new TaskId("foobarbaz", 1234567890), () -> false);
+    }
+
+    public TomlFoldTests(int num, EqlFoldSpec spec) {
+        this.num = num;
+        this.spec = spec;
+    }
+
+    private void testWithConfig(EqlConfiguration testConfig, EqlFoldSpec spec) {
+        Expression expr = parser.createExpression(spec.expression());
+
+        // resolve all function calls
+        expr = expr.transformUp(e -> {
+            if (e instanceof UnresolvedFunction) {
+                UnresolvedFunction uf = (UnresolvedFunction) e;
+
+                if (uf.analyzed()) {
+                    return uf;
+                }
+
+                String name = uf.name();
+
+                if (uf.childrenResolved() == false) {
+                    return uf;
+                }
+
+                String functionName = functionRegistry.resolveAlias(name);
+                if (functionRegistry.functionExists(functionName) == false) {
+                    return uf.missing(functionName, functionRegistry.listFunctions());
+                }
+                FunctionDefinition def = functionRegistry.resolveFunction(functionName);
+                Function f = uf.buildResolved(testConfig, def);
+                return f;
+            }
+            return e;
+        });
```

Comment:
This is essentially the `Analyzer.ResolveFunctions` rule. Instead of copying it, it's better to formalize it, especially since it exists and is pretty similar across EQL and SQL (so it can be parameterized and shared across projects /cc @astefan).
In testing, the expression would be wrapped in a basic/dummy `LogicalPlan` and the Analyzer or rule called directly on it.
In fact, since we have the `LogicalPlan`, the whole chain up to the Optimizer can be called directly as that will take care of the resolution and folding automatically and requires minimum setup in the tests. The advantage of using the optimizer is that the surrogate functions should be replaced as well automatically now and in the future.
An empty relationship should be enough considering everything is a literal.

Reply:
I'm fine with not blocking the current PR as long as there's a PR that indicates the follow-up work so this doesn't get lost.

Reply:
> In testing, the expression would be wrapped in a basic/dummy LogicalPlan and the Analyzer or rule called directly on it.

yeah, I wasn't exactly sure the best way to do that. I didn't want to pass it into a `Filter` because, I think that requires the type to be a boolean, and I wanted to validiate arbitrary expressions. Should I just create a dummy class for the testing here or use a different one? Put the expression within an instance of that class, analyze/optimize it, then get the optimized expression? That sounds fine to me.

I don't mind fixing that in this PR before merge

Reply:
You need to wrap the expression inside an existing plan - it can be a filter (which would need to be a bool so would need a comparison as well) or a projection (which would require an alias to be placed on top of the expression).
That should be enough.
Instead of creating a custom class which is heavier (code-wise and conceptually), it's easier to simply do the wrapping inside an utility method/write when you create the expression:

```
new Projection(Source.EMPTY, asSingletonList(new Alias("test", myExpression)));
```